locator/topology.cc

```diff
@@ -20,81 +20,308 @@ namespace locator {
+        if (!inserted_idx && iit->second != node.get()) {
+            if (node->host_id()) {
+                _nodes_by_host_id.erase(node->host_id());
+            }
+            if (node->endpoint() != inet_address{}) {
+                _nodes_by_endpoint.erase(node->endpoint());
+            }
+            on_internal_error(tlogger, format("Node idx already mapped: host_id={} endpoint={} idx={} dc={} rack={} state={}: currently mapped to host_id={} state={}",
+                    node->host_id(), node->endpoint(), node->idx(), node->dc_rack().dc, node->dc_rack().rack, node->get_state(),
+                    iit->second->host_id(), iit->second->get_state()));
+        }
 
-    if (current != _current_locations.end()) {
-        if (current->second.dc == dr.dc && current->second.rack == dr.rack) {
-            return;
+        const auto& dc = node->dc_rack().dc;
+        const auto& rack = node->dc_rack().rack;
+        const auto& endpoint = node->endpoint();
+        _dc_nodes[dc].emplace(node.get());
+        _dc_rack_nodes[dc][rack].emplace(node.get());
+        _dc_endpoints[dc].insert(endpoint);
+        _dc_racks[dc][rack].insert(endpoint);
+        _datacenters.insert(dc);
+
+        if (node->is_local()) {
+            _local_node = node;
         }
-        remove_endpoint(ep);
+        _all_nodes.emplace(node);
+    } catch (...) {
+        do_remove_node(node);
+        throw;
     }
+    return node;
+}
 
-    tlogger.debug("update_endpoint: {} {}/{}", ep, dr.dc, dr.rack);
-    _dc_endpoints[dr.dc].insert(ep);
-    _dc_racks[dr.dc][dr.rack].insert(ep);
-    _datacenters.insert(dr.dc);
-    _current_locations[ep] = std::move(dr);
+topology::mutable_node_ptr topology::make_mutable(const node_ptr& nptr) {
+    return const_cast<class node*>(nptr.get())->shared_from_this();
 }
 
-void topology::remove_endpoint(inet_address ep)
-{
-    auto cur_dc_rack = _current_locations.find(ep);
+node_ptr topology::update_node(node_ptr node, std::optional<host_id> opt_id, std::optional<inet_address> opt_ep, std::optional<endpoint_dc_rack> opt_dr, std::optional<node::state> opt_st) {
+    tlogger.debug("topology[{}]: update_node: node={} host_id={} endpoint={} dc={} rack={} state={}, at {}", fmt::ptr(this), fmt::ptr(node.get()),
+            opt_id.value_or(host_id::create_null_id()), opt_ep.value_or(inet_address{}), opt_dr.value_or(endpoint_dc_rack{}).dc, opt_dr.value_or(endpoint_dc_rack{}).rack, opt_st.value_or(node::state::none),
+            current_backtrace());
+    bool changed = false;
+    if (opt_id) {
+        if (*opt_id != node->host_id()) {
+            // FIXME: allow updating host_id for replace node.
+            // if (node->host_id()) {
+            //    on_internal_error(tlogger, format("Updating non-null node host_id is disallowed: host_id={} endpoint={}: new host_id={}",
+            //            node->host_id(), node->endpoint(), *opt_id));
+            // }
```

Comment:
note to self, this can be gone, it's now irrelevant and should never be relevant ever again.